File: smoke_tests/php/open_redirect.php
<?php
// Open Redirect vulnerabilities in PHP

// Test 1: Direct redirect from parameter
function redirect_unsafe() {
    $url = $_GET['url'];
    // VULNERABLE: Unvalidated redirect
    header("Location: $url");
    exit;
}

// Test 2: Redirect after login
function login_redirect() {
    // Authenticate user...
    $return_url = $_GET['return'];
    // VULNERABLE: Can redirect to external site
    header("Location: $return_url");
    exit;
}

// Test 3: Partial validation bypass
function safe_redirect() {
    $url = $_GET['url'];
    // VULNERABLE: Can bypass with //evil.com
    if (strpos($url, '/') === 0) {
        header("Location: $url");
        exit;
    }
}

// Test 4: Cookie-based redirect
function cookie_redirect() {
    $target = $_COOKIE['redirect_target'];
    // VULNERABLE: Cookie can be manipulated
    if (!empty($target)) {
        header("Location: $target");
        exit;
    }
}

// Test 5: Header-based redirect
function referer_redirect() {
    $referer = $_SERVER['HTTP_X_RETURN_URL'];
    // VULNERABLE: Header from client
    if ($referer) {
        header("Location: $referer");
        exit;
    }
}

// Test 6: Domain validation bypass
function domain_check() {
    $url = $_GET['url'];
    // VULNERABLE: evil.example.com contains example.com
    if (strpos($url, 'example.com') !== false) {
        header("Location: $url");
        exit;
    }
}

// Test 7: JavaScript redirect
function js_redirect() {
    $url = $_GET['url'];
    // VULNERABLE: JavaScript redirect
    echo "<script>window.location='{$url}';</script>";
}

// Test 8: Meta refresh redirect
function meta_redirect() {
    $target = $_GET['target'];
    // VULNERABLE: Meta refresh with user URL
    echo "<meta http-equiv='refresh' content='0;url={$target}'>";
}

// Test 9: URL from database
function dynamic_redirect() {
    $name = $_GET['name'];
    // Assume $url comes from database based on user input
    $url = get_redirect_url_from_db($name);
    // VULNERABLE: If database value came from user input
    header("Location: $url");
    exit;
}

// Test 10: PHP header with newlines
function header_injection() {
    $url = $_GET['url'];
    // VULNERABLE: URL can contain CRLF
    header("Location: " . $url);
    exit;
}

// Test 11: URL encoding bypass
function encoded_redirect() {
    $url = urldecode($_GET['url']);
    // VULNERABLE: Double encoding bypass
    if (strpos($url, '/') === 0) {
        header("Location: $url");
        exit;
    }
}

// Test 12: WordPress-style redirect
function wp_redirect() {
    $location = $_GET['redirect_to'];
    // VULNERABLE: Like wp_redirect without validation
    if (!empty($location)) {
        header("Location: $location", true, 302);
        exit;
    }
}

function get_redirect_url_from_db($name) {
    // Placeholder
    return '/default';
}
?>

File: smoke_tests/ruby/open_redirect.rb
# Open Redirect vulnerabilities in Ruby/Rails
class RedirectController < ApplicationController
  # Test 1: Direct redirect from params
  def login_redirect
    return_url = params[:return_url]
    # VULNERABLE: Unvalidated redirect
    redirect_to return_url
  end

  # Test 2: Redirect after authentication
  def after_login
    target = params[:target]
    # VULNERABLE: Can redirect to external site
    redirect_to target and return if target.present?
    redirect_to root_path
  end

  # Test 3: Partial validation bypass
  def safe_redirect
    url = params[:url]
    # VULNERABLE: Can be bypassed with //evil.com
    if url.start_with?('/')
      redirect_to url
    else
      redirect_to root_path
    end
  end

  # Test 4: Session-based redirect
  def restore_location
    return_url = session[:return_url]
    # VULNERABLE: Session value can be manipulated
    redirect_to return_url if return_url
  end

  # Test 5: Cookie-based redirect
  def cookie_redirect
    target = cookies[:redirect_target]
    # VULNERABLE: Cookie can be manipulated
    redirect_to target if target.present?
  end

  # Test 6: Header-based redirect
  def referer_redirect
    referer = request.headers['X-Return-Url']
    # VULNERABLE: Header from client
    redirect_to referer if referer
  end

  # Test 7: Domain validation bypass
  def domain_check
    url = params[:url]
    # VULNERABLE: evil.example.com contains example.com
    if url.include?('example.com')
      redirect_to url
    else
      redirect_to root_path
    end
  end

  # Test 8: JavaScript redirect
  def js_redirect
    url = params[:url]
    # VULNERABLE: JavaScript redirect
    render html: "<script>window.location='#{url}';</script>".html_safe
  end

  # Test 9: Meta refresh redirect
  def meta_redirect
    target = params[:target]
    # VULNERABLE: Meta refresh with user URL
    render html: "<meta http-equiv='refresh' content='0;url=#{target}'>".html_safe
  end

  # Test 10: URL from database
  def dynamic_redirect
    redirect_config = RedirectConfig.find_by(name: params[:name])
    # VULNERABLE: If database value came from user input
    redirect_to redirect_config.url
  end

  # Test 11: Redirect with path traversal
  def path_redirect
    path = params[:path]
    # VULNERABLE: Can add protocol with //
    redirect_to "/app/#{path}"
  end
end

# Sinatra example
require 'sinatra'
get '/redirect' do
  # VULNERABLE: Direct redirect
  redirect params[:url]
end
